src/storage/mod.rs

```diff
@@ -236,6 +247,20 @@ impl<E: Engine, L: LockManager, P: PdClient + 'static> Storage<E, L, P> {
                 // The bypass_locks set will be checked at most once. `TsSet::vec` is more efficient
                 // here.
                 let bypass_locks = TsSet::vec_from_u64s(ctx.take_resolved_locks());
+
+                // Update max_read_ts and check the in-memory lock table before getting the snapshot
+                if start_ts != TimeStamp::max() {
+                    concurrency_manager.update_max_read_ts(start_ts);
+                }
+                if ctx.get_isolation_level() == IsolationLevel::Si {
+                    concurrency_manager
+                        .read_key_check(&key, |lock| {
+                            lock.clone()
```

Comment:
Better to avoid cloning.

Reply:
I change `check_ts_conflict` to receive a `Cow<Lock>`.